Checkout Pull Request

smashwilson opened the pull request:
Build on the foundation established in #1556 to add pull request checkout functionality to the IssueishPaneItem.

##### Screenshots

| Before | After |
| -- | -- |
| <img width="350" alt="screen shot 2018-07-05 at 9 19 13 am" src="https://user-images.githubusercontent.com/17565/42325544-89a77c86-8034-11e8-8ec8-a00fe2ef7c2d.png"> | _TBD_ |

##### Remaining Work

- [x] Associate a working directory with each IssueishPaneItem by its URI. Use a WorkdirContextPool to identify the corresponding Repository.
- [x] Observe the Repository model in the IssueishDetailContainer and extract repository data for the IssueishDetailController.
- [ ] Pass a WorkdirContextPool to each IssueishPaneItem from the RootController that creates it.
- [ ] Introduce a MaybeOperation model to capture the common pattern of an action handler that (a) may be either enabled or disabled for various reasons and (b) toggles component state to `true` while performing some asynchronous work.
- [ ] Return an indexed `RemoteSet` from `repository.getRemotes()`, like we do with the `BranchSet` returned by `getBranches()`.
  - [ ] Update other callers of `getRemotes()` to use the `RemoteSet`.
- [ ] Create a MaybeOperation for the checkout action. Populate its enablement based on the relationship of the current issueish and the state of the repository.
- [ ] Write the handler method for the checkout action.
  - [ ] "Happy path:" fetch and checkout the PR's branch to a new branch in the local repository.
  - [ ] Handle the case where the PR already corresponds to a branch in the local repository.
  - [ ] Do something for the edge cases where a PR already corresponds to _multiple_ branches in the local repository.
  - [ ] Gracefully handle the edge case where the automatically generated local branch name already exists.
- [ ] Render a button in the IssueishDetailView that triggers the checkout action. Disable it and modify its text based on the operation state.
- [ ] Track changes to the current issueish by switching the active repository.
- [ ]  🚧 _doot doot doot_ 🚧 

##### Reviews

Before we merge, I'd like:

* A ✅ review from @simurai once he's happy with the design and UX flow.
* A ✅ review from @sguthals for Q&A

simurai commented on lib/controllers/issueish-detail-controller.js:
```diff
@@ -44,14 +68,112 @@ export class BareIssueishDetailController extends React.Component {
+    await this.props.fetch(fullHeadRef, {remoteName: sourceRemoteName});
+
+    // Check out the local ref and set it up to track the head ref.
+    await this.props.checkout(`pr-${issueish.number}/${issueish.headRefName}`, {
```
GitHub Desktop creates a branch name like `pr/${issueish.number}`. Should we do the same?

Would be a bit shorter. And also avoid showing the word "master" (like `pr-123/master`) in the status-bar when forks make changes on master and then submit a PR to the upstream repo.

smashwilson replied:
Interesting. I appended the head ref name mostly so that you'd get to see the entertaining branch names that your teammates come up with :wink:

simurai replied:
> to see the entertaining branch names that your teammates come up with

:smile:

Ok, after more testing in Desktop, it seems that:

- If the PR is _not_ coming from a fork, the local branch name is: `${issueish.headRefName}`
- If the PR _is_ coming from a fork, the local branch name is: `pr/${issueish.number}`

The command line instructions on .com:

- If the PR is _not_ coming from a fork, the local branch name is: `${issueish.headRefName}`
- If the PR _is_ coming from a fork, the local branch name is: `${username}-${issueish.headRefName}`

Hmm.. Not sure what's the best naming scheme, but interesting to see all the differences.

simurai commented on lib/controllers/issueish-detail-controller.js:
```diff
@@ -44,14 +68,112 @@ export class BareIssueishDetailController extends React.Component {
+      headPush.getShortRemoteRef() === issueish.headRefName;
+
+    if (fromPullRefspec || fromHeadRepo) {
+      return this.checkoutOp.disable(checkoutStates.CURRENT, 'Current');
```
Should it say what it is? Like `Current pull request`.

![image](https://user-images.githubusercontent.com/378023/42431833-b00b453c-8382-11e8-999d-c82ccd6ee939.png)


Or `Checked out`:

![image](https://user-images.githubusercontent.com/378023/42431822-9bf1a9ce-8382-11e8-8544-b28a755afd79.png)

smashwilson replied:
:+1: I went with "Checked out" to be a little less jarring on the transition from "Current pull request." Happy to be talked into any other copy though :smile:

I've also put the other status-dependent text in a button title rather than on the button text directly - the button will read "Checkout" unless you already have the PR's branch checked out, in which case it will say "Checked out." My logic is that the first time you see it, it would be confusing to have this big disabled button labelled "Rebase in progress," because you'd have no way to know what operation you couldn't perform because a rebase is in progress... on the other hand, I'm not sure how many people would notice the hover text as a way to figure out why it's disabled at a particular point in time.

smashwilson commented on lib/controllers/issueish-detail-controller.js:
```diff
@@ -44,14 +71,124 @@ export class BareIssueishDetailController extends React.Component {
+    return this.checkoutOp.enable();
+  }
+
+  async checkout() {
```
@jcansdale, @grokys, @sguthals: I'd be curious to hear if I made different tradeoffs here than you did in Visual Studio, and what your reasoning was if so. I waffled between a few different implementations before settling on this one.

Essentially:

1. I'm fetching the PR ref from the head repository, not the magic `refs/pulls/123/head` ref that dotcom gives you.
2. I'm adding a new named remote for each unique head repository. The remote name is the owner of the head repository, so checking out a PR from `smashwilson/atom` to `atom/atom` will give you a remote called `smashwilson`. (If a remote called `smashwilson` already exists and doesn't point to that repo, it'll fail.)
3. If your local repo already has a branch (with any name) that tracks the head ref from the head repo remote, it'll be checked out (and pulled with `--ff-only`).
4. Otherwise, the new head ref is fetched and checked out as a branch called `pr-123/head-ref-name`.

My priorities were:

* To avoid creating more and more branches if you checkout a PR that you've already checked out.
* To have `git fetch` and `git pull` work as expected (meaning, update with new work from the head ref) while you have the PR branch checked out, even if you run them from another client.
* To let you see the hilarious branch names contributors or your teammates come up with :trollface:.

The big downside to the way I've written it is that you'll slowly accumulate junk remotes in your config, especially if you're an open-source maintainer and you test PRs from different contributors often. But I don't think there's a way to automatically clean them up, and if you fetch from the head URL directly you can't set up a tracking relationship.

grokys replied:
Yes, they all sound like a good ideas. We made a few mistakes here (which I'm not going to list :P) and we need to re-do out checking out of PRs. This sounds like a more sensible approach, so if it looks like it works for you then we should move to doing this I think.

Do you agree @jcansdale?

> The big downside to the way I've written it is that you'll slowly accumulate junk remotes in your config

The alternative to this is instead of creating named remotes, put the remote URL in the branch config:

```
[branch "pr-123/foo"]  
    remote = https://github.com/smashwilson/atom
```

However we couldn't do this as libgit2sharp doesn't support it. If you're shelling out to `git` this should work.

smashwilson replied:
> [..] put the remote URL in the branch config [..]

Ahhh ha! Thank you, this is what I was looking for 🙇I _thought_ this was possible, but I kept looking for ways to do it with the `git checkout -b` or `git branch` commands rather than just setting the config keys myself. (`--track` verifies that its argument corresponds to an actual remote.)

jcansdale replied:
@smashwilson,

> Otherwise, the new head ref is fetched and checked out as a branch called `pr-123/head-ref-name`.

The problem with this scheme (and the one we use in GHfVS) is that there's no indication that the branch is tracking a remote for a different fork.

I'd be inclined to use something along the lines of: `pr-123/owner/head-ref-name`. That way when one of my PRs is checked out from parent repository, the branch would be named something like this: `pr-123/jcansdale/fixes/666-nasty-bug` (where `fixes/666-nasty-bug` is my branch name). That would make it easy to see at a glance where branches are coming from.

Do you prefer `pr-123` over `pr/123`?

In GHfVS we use `pr/123-pull-request-title` (title not branch name). 😆

jcansdale replied:
@grokys, 

> Yes, they all sound like a good ideas. We made a few mistakes here (which I'm not going to list :P) and we need to re-do our checking out of PRs. This sounds like a more sensible approach, so if it looks like it works for you then we should move to doing this I think.

Pretty much. I think one of our mistakes was not including the owner name. A lot of people don't realize that PR branches are literally branches from another fork. We should make this more obvious.

> If your local repo already has a branch (with any name) that tracks the head ref from the head repo remote, it'll be checked out (and pulled with --ff-only).

Yup, we should be doing this.

grokys replied:
> The problem with this scheme (and the one we use in GHfVS) is that there's no indication that the branch is tracking a remote for a different fork.

I don't think that's particularly a problem, personally. The real thing we did wrong there IMO was trying to convert the PR title into a branch name, which results in stupidly long branch names at times.

smashwilson replied:
> I'd be inclined to use something along the lines of: pr-123/owner/head-ref-name. That way when one of my PRs is checked out from parent repository, the branch would be named something like this: pr-123/jcansdale/fixes/666-nasty-bug (where fixes/666-nasty-bug is my branch name). That would make it easy to see at a glance where branches are coming from.

👍 I really like this. There's a nice symmetry with the way git names remote tracking branches there, too.

> I don't think that's particularly a problem, personally.

Maybe not! That's likely something that we could surface with the appropriate UXR, maybe? In any case, I don't think it'd _hurt_.

> Do you prefer pr-123 over pr/123?

Kind of? I _think_ that one's a purely aesthetic concern though. There might be implications for ease of scripting against them but I feel like anything you write there is already going to be using a bunch of `sed`-fu either way.

grokys replied:
> In any case, I don't think it'd hurt.

Yeah, I probably worded this badly: what I should have said is perhaps "I don't think this is the biggest problem with our naming scheme" - the big mistake we made was turning the PR title into the branch name.

jcansdale replied:
> 👍 I really like this. There's a nice symmetry with the way git names remote tracking branches there, too.

I'm all about the symmetry. ⚛️ 

That's why I was also wondering about:
```
pulls/123-jcansdale/fixes/666-nasty-bug
```

That would keep my `heads` folder nice and organized.

![image](https://user-images.githubusercontent.com/11719160/42577969-6f71c694-851d-11e8-9b49-9bcd721fc1f8.png)

That way `remotes`, `tags`, `pulls` and `fixes` would all describe the elements in the contained namespace (if that makes sense).

smashwilson commented:
> I might redesign the whole pane a bit, but for this PR, the button looks good. 👍

Looking forward to it 😁 

Where did you want to coordinate that, btw? I'd love to see what you have in mind. An issue? A PR to RFC-002?

smashwilson commented:
@sguthals: I think this is ready for a Q&A review whenever you get the chance :tada:

simurai commented:
> Where did you want to coordinate that, btw? I'd love to see what you have in mind. An issue? A PR to RFC-002?

Maybe just a PR with smaller changes and see how that feels. For things like PR reviews and diff, we can make RFCs and update the existing ones.

annthurium commented on lib/controllers/issueish-detail-controller.js:
```diff
@@ -44,14 +71,124 @@ export class BareIssueishDetailController extends React.Component {
+      return this.checkoutOp.disable(checkoutStates.HIDDEN, 'Cannot check out an issue');
+    }
+
+    if (this.props.isAbsent) {
```
would it maybe make sense to use some sort of enum prop here so you're not passing allll the props?  And then you could use a switch statement? (I don't feel super strongly, just a suggestion)

smashwilson replied:
Hmm you're right that it's awkward to query a bunch of `isXyz` props here. There's an enum-ish pattern that I use sometimes that might be nice for repository state branching like this:

```js
return this.props.repository.inState({
  absent: () => this.checkoutOp.disable(checkoutStates.HIDDEN, 'No repository found'),
  loading: () => this.checkoutOp.disable(checkoutStates.DISABLED, 'Loading'),
  // ...
  default: () => {},
});
```

Another alternative would be to pull more of this logic into the active Repository state:

```js
return this.props.repository.checkoutOpEnablement(this.checkoutOp);

// lib/models/repository-states/absent.js
export default class Absent extends State {
  checkoutOpEnablement(checkoutOp) {
    return checkoutOp.disable(checkoutStates.DISABLED, 'No repository found');
  }
}

// lib/models/repository-states/loading.js
export default class Loading extends State {
  checkoutOpEnablement(checkoutOp) {
    return checkoutOp.disable(checkoutStates.DISABLED, 'Loading');
  }
}
```

There are a few places that we do this already, though, so I think I'd rather address it in a code-cleanup PR rather than here, to be sure that we can come up with something that works nicely across the codebase 😄

annthurium commented on lib/controllers/issueish-detail-controller.js:
```diff
@@ -44,14 +71,124 @@ export class BareIssueishDetailController extends React.Component {
+
+    let sourceRemoteName, localRefName;
+
+    // Discover or create a remote pointing to the repo containing the pull request's head ref.
```
thank you for clarifying this with a comment!

annthurium commented on test/controllers/github-tab-controller.test.js:
```diff
@@ -0,0 +1,99 @@
+import React from 'react';
```
we didn't have any tests for github-tab-controller before?  yikes!  Thanks for adding some.

smashwilson replied:
Not a one 😅 

I was tempted to get full coverage on it here, but I decided not to because "refactoring it apart into `GitHubTabItem` / `GitHubTabContainer` / `GitHubTabController`" is on my radar at some point, so I figured I'd backfill the rest then.

annthurium commented:
@smashwilson overall, looks great!  Left a few comments.

One question: did you want to add metrics instrumentation? Like maybe clicks on the "checkout" button? I didn't see any but perhaps I'm missing something.

smashwilson commented:
> One question: did you want to add metrics instrumentation? Like maybe clicks on the "checkout" button? I didn't see any but perhaps I'm missing something.

The bulk of the work predated that PR getting merged, so you didn't miss any, there isn't any yet. But I ❤️ that idea - I'll add it to the checklist 👍

smashwilson commented:
> [..] put the remote URL in the branch config [..]

A bit of a hitch with this: it prevents the remote from appearing in the output of `for-each-ref`, which I'm using to identify which branch corresponds to which PR:

```
smashwilson @ hubtop ~/src/smashwilson/etch (pr-68/zaygraveyard/zkh-controlled-input)
$ git for-each-ref refs/heads/** --format '%(refname) - %(upstream) - %(upstream:remotename) - %(upstream:remoteref)'
refs/heads/master - refs/remotes/origin/master - origin - refs/heads/master
refs/heads/pr-67/zaygraveyard/zkh-false-child -  -  -
refs/heads/pr-68/zaygraveyard/zkh-controlled-input -  -  -
```

I'm going to see if I can get that information in a different way.

smashwilson commented:
Yah, sadly I don't think that'll work for me, either; fetch and pull work, but a bunch of git's other internals seem to rely on `branch.<name>.remote` being the name of a remote rather than a URL directly. 😞

smashwilson commented:
Metrics are in - I'm incrementing a `checkout-pr` counter each time that a PR is successfully checked out.

smashwilson commented:
I've added somewhat nicer error messages for anticipated git operation failures.

I wanted to make the errors a bit more specialized for the "Checkout PR" action, but at the moment our git operation error handling is centralized in a pipeline that's shared from everywhere that a Repository method is called. If I mentioned PR checkout specifically in the errors I added, we'd put ourselves in an awkward place if we call those git methods from anywhere else in the codebase.

Here's the verbiage I've settled on:

:star: _If you already have a remote named after the head repository owner that didn't point to the head repository on GitHub:_

<img width="555" alt="screen shot 2018-07-11 at 10 43 28 am" src="https://user-images.githubusercontent.com/17565/42580141-e2dfb0e4-84f7-11e8-9a2c-e490042cbb79.png">

:star: _If your existing local branch has diverged from the fetched head ref_

<img width="542" alt="screen shot 2018-07-11 at 10 43 59 am" src="https://user-images.githubusercontent.com/17565/42580312-348d0a9a-84f8-11e8-80b6-34f28b419d97.png">

sguthals commented:
👋 Ok some thoughts. 

First of all - really impressive work @simurai and @smashwilson!!! I'm really excited about this feature!

Happy to jump on a call to discuss any and all of these. 

--

<img width="724" alt="screen shot 2018-07-12 at 2 25 23 pm" src="https://user-images.githubusercontent.com/1314285/42660952-843a3c22-85e1-11e8-9264-17ff3eb83e43.png">
When we have multiple authors on a commit, this is how the UI looks. What if there are 10 contributors? Also, everything is right aligned, but then at the bottom in the separated section that avatar and text isn't also scooted over. I am OK with this design, but I want to make sure it was deliberate and not coincidental. Maybe more of a question for @simurai 

--

I clicked `Checkout` on this PR: https://github.com/atom/atom/pull/17140 and received this error:
<img width="1318" alt="screen shot 2018-07-12 at 2 43 23 pm" src="https://user-images.githubusercontent.com/1314285/42661097-028162f4-85e2-11e8-854f-348f8fc01c49.png">
I'm confused by the error message, as in, what action should I take to properly checkout this PR then? If it already exists, shouldn't I just switch to it then? Why wouldn't our functionality switch to it if it already existed?

Actually, this might be a bug? Because now I'm seeing that when I switch to branch pr-17681 the UI automatically updates with that being my current PR and "Checked Out", but when I switch to branch pr-17140 that doesn't happen. Why would that be? What would cause me to get into that state?

This is the state in the GitHub tab when I'm on that same branch:
<img width="404" alt="screen shot 2018-07-12 at 3 00 31 pm" src="https://user-images.githubusercontent.com/1314285/42661885-72d95334-85e4-11e8-8ab6-4c6df7496476.png">

Ah - I think I see, I think it's when I open a PR that is from a forked repo, because the same thing happens on this PR:
https://github.com/atom/atom/pull/17628

The messaging issue might be the same for the one you put above:
<img width="1318" alt="screen shot 2018-07-12 at 2 43 23 pm" src="https://user-images.githubusercontent.com/17565/42580141-e2dfb0e4-84f7-11e8-9a2c-e490042cbb79.png">

I think we might need to suggest actions for folks to follow to get to the state they want to get to, like you did here:
<img width="1318" alt="screen shot 2018-07-12 at 2 43 23 pm" src="https://user-images.githubusercontent.com/17565/42580312-348d0a9a-84f8-11e8-80b6-34f28b419d97.png">

-- 

Do we want reactions to look like this:
<img width="739" alt="screen shot 2018-07-12 at 2 43 02 pm" src="https://user-images.githubusercontent.com/1314285/42661237-71ff5a50-85e2-11e8-847a-1d5fd579dfec.png">
I can see how that matches the rest of the aesthetic, but for some reason looks really flat to me? It's also a little strange because I can't click on them, yet they look kind of like buttons? I like the idea of having them, but would it better to wait until we have support for them too?

--

Everything here at the top:
<img width="756" alt="screen shot 2018-07-12 at 2 48 37 pm" src="https://user-images.githubusercontent.com/1314285/42661318-b293c60a-85e2-11e8-98ef-f25244712bd7.png">
Is clickable except for the green `open` (which makes sense) and the title. The title, for some reason, threw me. I expected to be able to click on it and have it take me to the PR on dotcom. I know that the file path there at the top takes me there, but maybe because I'm used to clicking on titles and I'm used to things in white being links? Similar to how the thing that is a link here is in white and everything else is in grey?
<img width="756" alt="screen shot 2018-07-12 at 2 50 10 pm" src="https://user-images.githubusercontent.com/1314285/42661384-e3aeb042-85e2-11e8-8cfb-3dd15e02a032.png">

--

Should we add a line here between the last PR and `more`:
<img width="446" alt="screen shot 2018-07-12 at 2 50 41 pm" src="https://user-images.githubusercontent.com/1314285/42661426-ff89656e-85e2-11e8-8d1f-aac4db07278b.png">
Otherwise it looks like the more might be for that single PR?

--

I tried opening this issue: https://github.com/atom/atom/issues/17683 with the `GitHub: Open Issue or Pull Request" and I got a blank file?
<img width="1240" alt="screen shot 2018-07-12 at 2 57 22 pm" src="https://user-images.githubusercontent.com/1314285/42661730-e642feca-85e3-11e8-9186-f62aa9af4354.png">

--

I tried opening this PR: https://github.com/atom/atom/pull/17674 with the `GitHub: Open Issue or Pull Request and I got blank file also:
<img width="1245" alt="screen shot 2018-07-12 at 2 58 02 pm" src="https://user-images.githubusercontent.com/1314285/42661782-0f8a0436-85e4-11e8-82c7-8be48e71a386.png">

--

When I looked at this PR:
https://github.com/atom/atom/pull/17141
And I hovered over the `@Rezayovan` I got this error:
```
/Users/sarahguthals/Documents/GitHub/github/lib/views/issueish-timeline-view.js:132 unrecognized timeline event type: RenamedTitleEvent
_react2.default.createElement.groupedEdges.map @ /Users/sarahguthals/Documents/GitHub/github/lib/views/issueish-timeline-view.js:132
/Users/sarahguthals/Documents/GitHub/github/lib/views/issueish-timeline-view.js:132 unrecognized timeline event type: LabeledEvent
_react2.default.createElement.groupedEdges.map @ /Users/sarahguthals/Documents/GitHub/github/lib/views/issueish-timeline-view.js:132
/Users/sarahguthals/Documents/GitHub/github/node_modules/react-relay/lib/ReactRelayQueryRenderer.js:217 Uncaught TypeError: Cannot read property 'unstable_internal' of undefined
    at fetchQueryAndComputeStateFromProps (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-relay/lib/ReactRelayQueryRenderer.js:217:51)
    at new ReactRelayQueryRenderer (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-relay/lib/ReactRelayQueryRenderer.js:77:8)
    at constructClassInstance (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:11447:18)
    at updateClassComponent (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:13144:7)
    at beginWork (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:13824:14)
    at performUnitOfWork (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:15863:12)
    at workLoop (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:15902:24)
    at HTMLUnknownElement.callCallback (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:100:14)
    at Object.invokeGuardedCallbackDev (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:138:16)
    at invokeGuardedCallback (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:187:29)
    at replayUnitOfWork (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:15310:5)
    at renderRoot (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:15962:11)
    at performWorkOnRoot (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16560:22)
    at performWork (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16482:7)
    at performSyncWork (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16454:3)
    at requestWork (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16354:5)
    at scheduleWork$1 (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16218:11)
    at scheduleRootUpdate (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16785:3)
    at updateContainerAtExpirationTime (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16812:10)
    at updateContainer (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16839:10)
    at ReactRoot.render (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:17122:3)
    at /Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:17262:14
    at unbatchedUpdates (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16679:10)
    at legacyRenderSubtreeIntoContainer (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:17258:5)
    at Object.render (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:17317:12)
    at UserMentionTooltipItem.get element [as element] (/Users/sarahguthals/Documents/GitHub/github/lib/items/user-mention-tooltip-item.js:48:34)
    at UserMentionTooltipItem.getElement (/Users/sarahguthals/Documents/GitHub/github/lib/items/user-mention-tooltip-item.js:14:12)
    at ViewRegistry.module.exports.ViewRegistry.createView (/private/var/folders/tk/mrlhdh_9507_4_nz58bz3mjr0000gn/T/AppTranslocation/8…A022B9C27DC/d/Atom.app/Contents/Resources/app.asar/src/view-registry.js:87:26)
    at ViewRegistry.module.exports.ViewRegistry.getView (/private/var/folders/tk/mrlhdh_9507_4_nz58bz3mjr0000gn/T/AppTranslocation/8…A022B9C27DC/d/Atom.app/Contents/Resources/app.asar/src/view-registry.js:75:21)
    at Tooltip.setContent (/private/var/folders/tk/mrlhdh_9507_4_nz58bz3mjr0000gn/T/AppTranslocation/8…309-BA022B9C27DC/d/Atom.app/Contents/Resources/app.asar/src/tooltip.js:342:41)
/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:14226 The above error occurred in the <ReactRelayQueryRenderer> component:
    in ReactRelayQueryRenderer

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://fb.me/react-error-boundaries to learn more about error boundaries.
logCapturedError @ /Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:14226
/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16543 Uncaught TypeError: Cannot read property 'unstable_internal' of undefined
    at fetchQueryAndComputeStateFromProps (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-relay/lib/ReactRelayQueryRenderer.js:217:51)
    at new ReactRelayQueryRenderer (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-relay/lib/ReactRelayQueryRenderer.js:77:8)
    at constructClassInstance (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:11447:18)
    at updateClassComponent (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:13144:7)
    at beginWork (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:13824:14)
    at performUnitOfWork (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:15863:12)
    at workLoop (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:15902:24)
    at renderRoot (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:15942:7)
    at performWorkOnRoot (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16560:22)
    at performWork (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16482:7)
    at performSyncWork (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16454:3)
    at requestWork (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16354:5)
    at scheduleWork$1 (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16218:11)
    at scheduleRootUpdate (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16785:3)
    at updateContainerAtExpirationTime (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16812:10)
    at updateContainer (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16839:10)
    at ReactRoot.render (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:17122:3)
    at /Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:17262:14
    at unbatchedUpdates (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:16679:10)
    at legacyRenderSubtreeIntoContainer (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:17258:5)
    at Object.render (/Users/sarahguthals/Documents/GitHub/github/node_modules/react-dom/cjs/react-dom.development.js:17317:12)
    at UserMentionTooltipItem.get element [as element] (/Users/sarahguthals/Documents/GitHub/github/lib/items/user-mention-tooltip-item.js:48:34)
    at UserMentionTooltipItem.getElement (/Users/sarahguthals/Documents/GitHub/github/lib/items/user-mention-tooltip-item.js:14:12)
    at ViewRegistry.module.exports.ViewRegistry.createView (/private/var/folders/tk/mrlhdh_9507_4_nz58bz3mjr0000gn/T/AppTranslocation/8…A022B9C27DC/d/Atom.app/Contents/Resources/app.asar/src/view-registry.js:87:26)
    at ViewRegistry.module.exports.ViewRegistry.getView (/private/var/folders/tk/mrlhdh_9507_4_nz58bz3mjr0000gn/T/AppTranslocation/8…A022B9C27DC/d/Atom.app/Contents/Resources/app.asar/src/view-registry.js:75:21)
    at Tooltip.setContent (/private/var/folders/tk/mrlhdh_9507_4_nz58bz3mjr0000gn/T/AppTranslocation/8…309-BA022B9C27DC/d/Atom.app/Contents/Resources/app.asar/src/tooltip.js:342:41)
    at Tooltip.show (/private/var/folders/tk/mrlhdh_9507_4_nz58bz3mjr0000gn/T/AppTranslocation/8…309-BA022B9C27DC/d/Atom.app/Contents/Resources/app.asar/src/tooltip.js:227:10)
    at Tooltip.enter (/private/var/folders/tk/mrlhdh_9507_4_nz58bz3mjr0000gn/T/AppTranslocation/8…309-BA022B9C27DC/d/Atom.app/Contents/Resources/app.asar/src/tooltip.js:178:17)
    at HTMLAnchorElement.innerHandler (/private/var/folders/tk/mrlhdh_9507_4_nz58bz3mjr0000gn/T/AppTranslocation/8…9C27DC/d/Atom.app/Contents/Resources/app.asar/src/delegated-listener.js:24:7)
```

Ok - looks like I'm getting that error every time I hover over a link that is in the description or comments of the PR. The links still work, however.

--

smashwilson commented:
@sguthals: wow, thanks for the thoroughness 😁

This PR only touched the Issueish pane to add the checkout button, but like @simurai said, he's been contemplating a comprehensive redesign of the rest of the pane soon. Hopefully he can take your feedback into account as he's tackling that 👍 

> I'm confused by the error message, as in, what action should I take to properly checkout this PR then? If it already exists, shouldn't I just switch to it then? Why wouldn't our functionality switch to it if it already existed?
> 
> Actually, this might be a bug? Because now I'm seeing that when I switch to branch pr-17681 the UI automatically updates with that being my current PR and "Checked Out", but when I switch to branch pr-17140 that doesn't happen. Why would that be? What would cause me to get into that state?

I'm confused, too - it should have detected that pr-17681 corresponded to the PR by its configured upstream. Can you run the following command in that repository and share its output:

```
git for-each-ref --format '%(upstream) - %(upstream:remotename) - %(upstream:remoteref) | %(push) - %(push:remotename) - %(push:remoteref)' refs/heads/pr-17140/Rezayovan/master
```

This could be something like a mismatch in the way the remote is set up or incorrectly cached data. Does it detect properly if you reload the Atom window?

> I think we might need to suggest actions for folks to follow to get to the state they want to get to, like you did here:

The problem here is that I can't currently customize the error messaging based on the originator of the `git checkout` call. So If I changed it to be more specific about the checkout PR action, it would show that for _any_ time you tried to check out a branch and got that error.

I have an idea about ways to improve our error handling on a per-callsite basis, but it's a bit of a sprawling change and would benefit other places too, so I was hoping to defer that to a separate PR.

> I tried opening this issue: atom/atom#17683 with the `GitHub: Open Issue or Pull Request" and I got a blank file?

Are you sure you're on the latest version of this branch? I thought I fixed that earlier this week. Can you verify the commit of atom/github you're running?

> Should we add a line here between the last PR and more: [...]

That sounds good to me, but it'd be out of scope for this PR. @simurai, any thoughts?

> Ok - looks like I'm getting that error every time I hover over a link that is in the description or comments of the PR. The links still work, however.

Oh weird. What kind of links - to another issue/pull request, an @-mention, or any hyperlink at all? I'll add a note to investigate this one.

sguthals commented:
@smashwilson -
Ran the command:
```
Sarahs-iMac:atom sarahguthals$ git for-each-ref --format '%(upstream) - %(upstream:remotename) - %(upstream:remoteref) | %(push) - %(push:remotename) - %(push:remoteref)' refs/heads/pr-17140/Rezayovan/master
 -  -  |  - https://github.com/Rezayovan/atom.git - 
Sarahs-iMac:atom sarahguthals$ 
```

Reloading did not help.

--

> The problem here is that I can't currently customize the error messaging based on the originator of the git checkout call. So If I changed it to be more specific about the checkout PR action, it would show that for any time you tried to check out a branch and got that error.
> I have an idea about ways to improve our error handling on a per-callsite basis, but it's a bit of a sprawling change and would benefit other places too, so I was hoping to defer that to a separate PR.

Got it! No problem 👍 

--

> Are you sure you're on the latest version of this branch? I thought I fixed that earlier this week. Can you verify the commit of atom/github you're running?

Ughhh I'm really sorry - you're right, I hadn't fetched. That is in fact working.

--
> Oh weird. What kind of links - to another issue/pull request, an @-mention, or any hyperlink at all? I'll add a note to investigate this one.

So it only pops up the console when I hover on an @ mention, but then when the console is open, I do see the error on any link

--

Out of Scope for this PR, but to keep it all in one place for right this second, I also noticed that there is no messaging around the fact that I have un-staged/ un-committed changes on a branch in the side panel:
<img width="402" alt="screen shot 2018-07-12 at 4 50 51 pm" src="https://user-images.githubusercontent.com/1314285/42665298-ddd97f24-85f3-11e8-9317-e3305dc271ee.png">

sguthals commented:
From your comment in Slack, you're right, after I pulled I now get this error still:
<img width="517" alt="screen shot 2018-07-12 at 4 53 22 pm" src="https://user-images.githubusercontent.com/1314285/42665359-25286264-85f4-11e8-8c6e-818536d7c4b1.png">

However, below the error I see:
<img width="399" alt="screen shot 2018-07-12 at 4 53 27 pm" src="https://user-images.githubusercontent.com/1314285/42665365-2ae1c51a-85f4-11e8-90f8-33874ca97477.png">

smashwilson commented:
> From your comment in Slack, you're right, after I pulled I now get this error still: [...]

Ah, right, you'd have to delete the branch and check out the PR again to get it configured properly. (Or you could try it with another PR from a fork?)

simurai commented:
> @sguthals Should we add a line here between the last PR and more: [...]
> @smashwilson That sounds good to me, but it'd be out of scope for this PR. @simurai, any thoughts?

Chatted with @sguthals earlier today and yeah, the plan is to make a new PR for all the visual improvements. 👍

smashwilson commented:
Hovers and clicks in rendered GitHub markdown work without a stack trace:

<img width="558" alt="screen shot 2018-07-16 at 8 58 18 am" src="https://user-images.githubusercontent.com/17565/42761768-754a669e-88dc-11e8-84bd-1ebd121550cc.png">

@sguthals: Ready for another pass ✨ 🔎
